### PR TITLE
Optimize startup time and memory usage

### PR DESCRIPTION
**Pull request by roman-khimov**
### Problem

Reading all of the hashes takes time, storing them requires a lot of memory.

### Solution

Only get the ones we need an only when they are needed.

### CODE REVIEW

**Review by AnnaShaleva — APPROVED** [review]

File: pkg/services/rpcsrv/server.go [hunk 1]
```diff
@@ -2681,16 +2678,16 @@
...
 	if err != nil {
 		return 0, neorpc.ErrInvalidParams
 	}
 
-	if num < 0 || num > int(s.chain.BlockHeight()) {
+	if num < 0 || uint32(num) > s.chain.BlockHeight() {
 		return 0, invalidBlockHeightError(0, num)
 	}
-	return num, nil
+	return uint32(num), nil
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/services/rpcsrv/server.go:2690]: We already have upper bound check (not higher than current block height) so we may add ignoring comment for CodeQL.

[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/services/rpcsrv/server.go:2690]: I've dismissed it already, shouldn't be a problem.

**Comment by roman-khimov** [comment]
Rebased this one.

File: pkg/core/blockchain.go [hunk 2]
```diff
@@ -721,7 +686,7 @@ func (bc *Blockchain) resetStateInternal(height uint32, stage stateChangeStage)
 			blocksCnt, batchCnt, keysCnt int
 		)
 		for i := height + 1; i <= currHeight; i++ {
-			err := cache.DeleteBlock(bc.GetHeaderHash(int(i)))
+			err := cache.DeleteBlock(bc.GetHeaderHash(i))
```
Comment (AnnaShaleva) [inline pkg/core/blockchain.go:689]: Nitpicking, this one belongs to a previous commit.

**Review by AnnaShaleva — APPROVED** [review]

File: pkg/core/blockchain.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/core/blockchain.go:689]: Fixed.

**Review by AnnaShaleva — APPROVED** [review]

**Pull request merged**